Support handling LogoutResponse from SAML idP

ywangd opened the pull request:
SAML idP sends back a LogoutResponse at the end of the logout workflow. It can be sent via either HTTP-Redirect binding or HTTP-POST binding. Currently, the HTTP-Redirect request is simply ignored by Kibana and never reaches ES. It does _not_ cause any obvious issue and the workflow is completed normally from user's perspective. For completeness, we have an open issue #40901 for handling it on the ES side.

The HTTP-POST request results in a `404` error because `POST` request is not accepted by Kibana's logout end-point. This causes a non-trivial issue because it renders an error page in user's browser. In addition, some resources do not seem to be fully cleaned up due to the error, e.g. the username will be pre-filled when trying to login again after the 404 error.

This PR tries to solve both of the above issues with a new `_security/saml/complete_logout` end-point. The function is largely a subset of the API that checks SAML authentication response.

NOTE: This PR is for elasticsearch only. To have complete end-to-end support for SAML logout response, changes need to be made for Kibana to handle (forward) the response message to elasticsearch.

Related to: #40901
Resolves: #43264

ywangd committed:
Initial working version. No tests yet

ywangd committed:
Refactor and add tests

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

ywangd commented:
This PR is not entirely done yet. But I'd appreciate some feedback on the overall approach since this is my first saml related PR. Thanks.

Still need: 
* Docs
* More tests
* API spec? (but we don't seem to have it other than for idP)
* yaml tests (again don't seem to have it?)

Changes are also needed on Kibana side to really make the flow work. I could create an issue in Kibana repo (a related issue is elastic/kibana#27156)?

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,46 @@
+            if (logoutResponse.isSigned()) {
+                validateSignature(logoutResponse.getSignature());
+            }
+            checkInResponseTo(logoutResponse, allowedSamlRequestIds);
```
This requires Kibana to know the `_ID` of the logout request crafted by ES and sent to the idP. The current `SamlLogoutResponse` does not give this information in a way that Kibana can easily access. 

We could either change `SamlLogoutResponse` to include a separate field of `_ID` similar to what we do with `SamlPrepareAuthenticationResponse`. Or we could simply skip this check?

jkakavas replied:
The spec ([4.4.4.1](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf)) doesn't talk about validating the `InResponseTo` parameter, but I've seen other libraries / impementation do that. In all honesty SAML SLO is such a wild wild west, everyone does pretty much what they feel like :/ 

Given that it's not dictated by the specification and that our API has a binary response of "The Logout Response is fine" vs "the Logout Response is _not_ fine" which cannot be abused somehow ( ie. open redirect etc) I think it would be fine to not validate

ywangd replied:
It's straightforward to add a new `requestId` field to the `SamlLogoutResponse` ES produces (via the `/_security/saml/logout` API). Given it is a change to an existing API, we may need to communicate to the Kibana team earlier on. @legrego Could you please comment on this? The `requestId` value is then required to be passed back to ES from Kibana to verify the `LogoutResponse` from the idP.

jkakavas replied:
> Given it is a change to an existing API, we may need to communicate to the Kibana team earlier on

In general, adding additional fields to our responses is not considered to be a breaking change so I believe Kibana will be fine with it even if we don't sync merging support for this. 

> The requestId value is then required to be passed back to ES from Kibana to verify the LogoutResponse from the idP.

Kibana does that already for the request ID of the authentication response and send this back to ES in the `_security/saml/authenticate` API so I believe it would be fine to do here too.

ywangd replied:
Thanks @jkakavas I'll make the change accordingly.

azasypkin replied:
> In general, adding additional fields to our responses is not considered to be a breaking change so I believe Kibana will be fine with it even if we don't sync merging support for this.

That's correct.

> Kibana does that already for the request ID of the authentication response and send this back to ES in the _security/saml/authenticate API so I believe it would be fine to do here too.

The main difference here is that request ID returned from `_security/saml/prepare` is stored in the cookie. But during logout flow we used to clear cookie when user is redirected to IdP with `LogoutRequest`. Having said that, we can change that, we'll just need to carefully fix all the remaining places that still treat _any_ cookie as a sign of authenticated user.

jkakavas replied:
Since the specification _doesn't_ explicitly call this out *AND* the outcome is in any case a `no-op`, I would be ok with not validating this. Especially since it sounds possible that the risk of introducing a bug on Kibana's side by changing this behavior is not minimal. On the other hand, treating "any cookie as a sign of authenticated user." is something worth fixing regardless of that change so if this is an extra driver for that change, let's do it !

ywangd committed:
fix copy/paste text

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,46 @@
+            if (logoutResponse == null) {
+                throw samlException("Cannot convert element {} into LogoutResponse object", root);
+            }
+            if (logoutResponse.isSigned()) {
```
The response _MUST_ be signed ([4.4.4.2](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf))  so we should fail if it is not signed.

ywangd replied:
The spec says "Must authenticate itself either by signing or a binding specific mechansim". While the HTTP-POST `LogoutResponse` is signed, the `LogoutResponse` from HTTP-Redirect binding is _not_. So I would prefer it to be optional.

jkakavas replied:
> or a binding specific mechanism

this alludes to the artifact binding that is not used/supported here

> . While the HTTP-POST LogoutResponse is signed, the LogoutResponse from HTTP-Redirect binding is not.

Why are you stating that as a fact ? It can very well be signed and according to the specification it `MUST` be.

ywangd replied:
> Why are you stating that as a fact ?
>
It's not an universal fact. I was trying to avoid explicitly mentioning a popular idP's name. Sorry about the confusion. If we mandate signature verification and Kibana starts to redirect the response to ES, the verification will fail since it does not have signature.

In addition, including signature in HTTP-Redirect request may risk hitting the (practical) limit of URL length.

jkakavas replied:
>  If we mandate signature verification and Kibana starts to redirect the response to ES, the verification will fail since it does not have signature.

It all boils down to how permissive we want to be with things that go against the specification. If a popular IDP doesn't adhere to the specification, do we happily consume what they are sending us or do we not ? And if another popular IDP comes along and doesn't adhere to the standards in a different way, do we happily accept that too? Where do we draw the line? I believe that drawing the line at what the specification says is easier to argue about than drawing the line in an arbitrary leniency we deem acceptable at some point. 

> In addition, including signature in HTTP-Redirect request may risk hitting the (practical) limit of URL length. I guess this is probably the reason why it is not signed by the aforementioned idP

I doubt that. SAML Logout Response messages are quite small ( comparable to authentication requests that also use the HTTP Redirect binding and are signed )

ywangd replied:
> I believe that drawing the line at what the specification says is easier to argue about than drawing the line in an arbitrary leniency we deem acceptable at some point.
> 
This I agree. But the whole story is not only just with a particular idP. The leniency was already implemented by us (by dropping the `LogoutResponse` at Kibana side). We could choose to tighten up the behaviour and accept the fact it might break user's existing setup. I am happy to conform the standard if we are certain that is the price we would like to pay.

> authentication requests that also use the HTTP Redirect binding and are signed
> 
The auth `Response` from idP is signed but uses HTTP-POST (in fact we only support HTTP-POST binding for this). Our SAML AuthRequest uses redirect and it is not signed. As far as I know, Apache server has 4096 (or is it 8192) limit on URL length, it is a thing when URL gets too long.

I have to admit that I only just started to read the spec and have many knownlege gaps. For an example, the signature part is confusing to me. It mandates a `LogoutResponse` to be signed, but signature for authn `Response` is optional. Isn't the authn `Response` a more sensitive/valuable information compared to `LogoutResponse`?

jkakavas replied:
> The leniency was already implemented by us (by dropping the LogoutResponse at Kibana side).

This was not leniency. This was part of the specification not handled. If we _do_ handle it, let's do it correctly.

> The auth Response from idP is signed but uses HTTP-POST (in fact we only support HTTP-POST binding for this)

True, but I said 

> authentication requests that also use the HTTP Redirect binding and are signed

and SAML Authentication Requests are _not_ SAML Authentication Responses :)

>  Our SAML AuthRequest uses redirect and it is not signed. 

It is when it is configured to be https://www.elastic.co/guide/en/elasticsearch/reference/master/saml-guide-authentication.html#_configuring_elasticsearch_for_signing

> As far as I know, Apache server has 4096 (or is it 8192) limit on URL length, it is a thing when URL gets too long.

And SAML with authentication requests ( using the HTTP Redirect binding) being signed has been out there for 15 years now, this is heavily supported and used, I don't think we need to worry about this. 

> but signature for authn Response is optional

I don't think this is true.

ywangd replied:
First of all, thanks for indulging me with the discussions. I really appreciate it. It helps me understanding the issue better ❤️ 
> It is when it is configured to be
>
Thanks for pointing this out. I completely ignored this part when setting up saml locally. But isn't it actually an example that we do allow unsigned `LogoutResponse` (the difference being we are the sender in this case)? 

> SAML with authentication requests ( using the HTTP Redirect binding) being signed has been out there for 15 years now, this is heavily supported and used
>
This is a good point as well. Given authn request can be signed and use HTTP Redirect, I agree there will be no issue for `LogoutResponse`.

> I don't think this is true.
>
You are right that it is not "optional". The word used in the spec [6.4.2](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf) is "SHOULD". It is stronger than optional, but not MUST. Our implementation for this part treats it as optional:
https://github.com/elastic/elasticsearch/blob/a443e1387dc262b3ec58b53c6b2f2001a781093a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticator.java#L90-L92

jkakavas replied:
> But isn't it actually an example that we do allow unsigned LogoutResponse (the difference being we are the sender in this case)?

I don't remember the discussions we had while implementing this , I can assume we made a decision that interoperability is more important than specification adherence at that time. We can revisit that decision too :) 


> You are right that it is not "optional". The word used in the spec 6.4.2 is "SHOULD".

We only support the web browser single sign on profile, so 4.1 in that spec, see 4.1.3.5


>  Our implementation for this part treats it as optional:

It -hopefully- doesn't!!! A SAML response contains SAML assertions. What we do support is either the response or the assertion is signed :) There must be an integrity protection implemented, otherwise it would be trivial for attackers to bypass authentication

ywangd replied:
> We only support the web browser single sign on profile, so 4.1 in that spec, see 4.1.3.5
> 
You are right. I read the wrong section 🤦 

> What we do support is either the response or the assertion is signed
> 
Yes that's what I figured as well. There is no risk in this part. 

I think the only thing that bothers me is this "popular idP". The original issue #43264 is created with it as an example of why we consider to support HTTP-POST LogoutResponse. So it feels rather unsatifying if we fix one binding and breaks the other (the current working one).

The code change itself either way is trivial. The decision making part is more important. I feel we could talk about this over sync time.

jkakavas replied:
> I think the only thing that bothers me is this "popular idP".

I don't think we need to make changes that satisfy any "popular IDP", we should make changes that make sense for the users and are correct. What's more, I see no indication whatsoever that any IDP doesn't support signed Logout Responses so I'm not sure what we would be actually breaking with this. 

I think we should not be lenient here for no reason, but happy to discuss this more on the sync.

ywangd replied:
Sync update: we decided to conform to the spec and perform full validation. Caller of the API can decide what to do with validation errors if any. Thanks for the discussion.

ywangd committed:
Add requestId to logout response for kibana to pass it back

ywangd committed:
checkstyle

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlLogoutResponse.java:
```diff
@@ -16,24 +16,32 @@
 
-    public SamlLogoutResponse(String redirectUrl) {
+    public SamlLogoutResponse(String requestId, String redirectUrl) {
+        this.requestId = requestId;
```
I don't think we need handle BWC here since this response never goes across nodes. The only consumer is Kibana (or other external system that integrates with ES).

ywangd committed:
Support LogoutResponse with both HTTP-Redirect and HTTP-Post bindings

ywangd commented:
Updated to ensure support of handling SAML LogoutResponse via both HTTP-Redirect and HTTP-Post bindings. The new `/_security/saml/verify_logout` API is used to handle both scenarios. In both cases, it takes a POST request with following JSON payload:
```json
{
  "ids": [...],
  "realm": "saml1",
  "content": "..."
}
```

where `ids` is a list of request IDs that `InResponseTo` value should be verified against, and `content` is the raw URI query string of HTTP-Redirect binding or raw body of HTTP-Post binding.

The code examines the `content` to see whether an URL level signature is available. If it is found, HTTP-Redirect is assumed and verified accordingly. If no URL level signature is found, HTTP-Post is assumed, which mandates the signature at XML level.

ywangd committed:
Rename and checkstyle

ywangd committed:
Forbidden API

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandlerHttpRedirectTests.java:
```diff
@@ -0,0 +1,132 @@
+        final String url = new SamlRedirect(logoutResponse, signingConfiguration).getRedirectUrl();
+        final ElasticsearchSecurityException e =
+            expectSamlException(() -> samlLogoutResponseHandler.handle(new URI(url).getRawQuery(), List.of(requestId)));
+        assertThat(e.getMessage(), containsString("Failed to parse SAML message"));
```
related to the discussion above, this is a case where we don't want to fail to parse the SAML message because we will try to parse it as if it came from the HTTP POST binding. We want to fail this because it is an unsigned request that came via the HTTP-Redirect binding and we disallow that.

ywangd replied:
I think I see your point here. I'll move away from auto-detect and have two separate fields at request level.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,59 @@
+    public void handle(String content, Collection<String> allowedSamlRequestIds) {
+        final ParsedQueryString parsed = parseQueryStringAndValidateSignature(content, "SAMLResponse");
+        final Element root;
+        if (parsed.hasSignature) {
```
I would very much prefer that we are explicit here and not try to deduce and imply checks. I think we need two parameters for the REST layer, one for the body of a response that comes via the HTTP-POST binding and a _different_ one for the query string of a response that comes via the HTTP-Redirect binding. Then we can be very explicit here with regards to what we expect to have and what we need to process/check/validate

ywangd replied:
See above.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlObjectHandler.java:
```diff
@@ -322,4 +331,74 @@ protected void validateNotOnOrAfter(DateTime notOnOrAfter) {
         }
     }
+
+    protected ParsedQueryString parseQueryStringAndValidateSignature(String queryString, String samlMessageParameterName) {
```
Again, assuming that you didn't change much when you copied this over ( apart from introducing `samlMessageParameterName` ) , please do flag if otherwise ! ( :octocat: diffs are not that helpful in spotting this kind of things :) )

ywangd replied:
Good spot! Indeed this is the only significant change as part of the refactoring. Next time I'll compile the list of changes before hand to save you some time. Sorry about that. Here is a list for rest of the changes (they are all very minor):
* method visibility from `private` to `protect` due to superclass extraction. 
  - This includes: `SamlResponseHandler#isSuccess`, `SamlResponseHandler#getStatusCodeMessage`, `SamlObjectHandler#decodeBase64`, `SamlObjectHandler#inflate`, and `SamlObjectHandler#parseQueryStringAndValidateSignature`.
* Reverse the order of String equals compare in `SamlResponseHandler#isSuccess`. 
  - Changed from `status.getStatusCode().getValue().equals(StatusCode.SUCCESS)` to `StatusCode.SUCCESS.equals(status.getStatusCode().getValue())`.

jkakavas reviewed (changes requested):
Looking good Yang! I added some comments , a couple of these I believe we should address, the rest are suggestions that we can iterate on

azasypkin commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlVerifyLogoutRequest.java:
```diff
@@ -0,0 +1,71 @@
+    @Nullable
+    private String realm;
+    @Nullable
+    private String assertionConsumerServiceURL;
```
Correct, [starting from 7.7](https://github.com/elastic/elasticsearch/pull/52188) we store `realm` returned from the `_security/saml/authenticate` in the cookie/session and we'll keep it when store Logout Request ID there as well.

ywangd replied:
Thanks. Good to know that. Removed.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlVerifyLogoutAction.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public final class SamlVerifyLogoutAction extends ActionType<SamlVerifyLogoutResponse> {
+
+    public static final String NAME = "cluster:admin/xpack/security/saml/verify_logout";
```
I was not particular happy about the name `verify_logout` either. Changed it to `complete_logout` and also for names of all relevant classes.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticator.java:
```diff
@@ -333,7 +277,13 @@ private void checkSubject(Subject assertionSubject, XMLObject parent, Collection
         checkRecipient(confirmationData.get(0));
         checkLifetimeRestrictions(confirmationData.get(0));
-        checkInResponseTo(confirmationData.get(0), allowedSamlRequestIds);
+        SubjectConfirmationData subjectConfirmationData = confirmationData.get(0);
```
Sure, named it `checkSubjectInResponseTo`. Let me know if you have a better name.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlVerifyLogoutRequest.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public final class SamlVerifyLogoutRequest extends ActionRequest {
+
+    private String content;
```
I was intentionally trying to make it the same to handle both bindings. The current approach is to have a single `POST` endpoint that Kibana can call for both HTTP-Redirect and HTTP-Post bindings from the idP, i.e. Kibana acts as a multiplexer here. So what ES receives is an uniform `POST` message. The handler (`SamlLogoutResponseHandler`) is able to tell whether the paylod (`content`) itself is signed and proceeds as HTTP-Redirect. If it is not, it proceeds as HTTP-POST and requires the XML to be signed, which is more or less similar to what `SamlAuthenticator` does. 

I can be convinced to have two separate endpoints, one for HTTP-Redirect and one for HTTP-POST. If we do that, we can even have two separate Request classes as well. But if we choose to have a single endpoint, I'd prefer to also not differentiate it at the Request object level.

Whether we should have one or two end-points also depends on what Kibana can do (@azasypkin). My thinking is that Kibana will have to prepare the requset to ES anyhow, i.e. it has to add the realm and requestId information. So Kibana might as well just act as the multiplexer and prepare an uniform message.

ywangd replied:
After reading your comment below. I am convinced to have separate fields at request level, `query` and `body` for HTTP-Redirect and HTTP-POST, respectively. The `SamlCompleteLogoutRequest` will validate that they cannot both be `null` nor both be set. It also has convenient methods, `getContent` and `isHttpRedirect`, so the information is clear at handler level. Thanks.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlResponseHandler.java:
```diff
@@ -0,0 +1,93 @@
+
+import static org.elasticsearch.xpack.security.authc.saml.SamlUtils.samlException;
+
+public class SamlResponseHandler extends SamlObjectHandler {
```
This is mainly because we do have a real Request object from idP, i.e.`SamlInvalidateSessionRequest`. Its associated handler is `SamlLogoutRequestHandler`, which is a subclass of `SamlObjectHandler`. In another word, `SamlObjectHandler` is the base class for both Request and Response handlers. So I did not merge `SamlResponseHandler` into `SamlObjectHandler`.

Also I did a few refactoring which is likely made the changes hard to track, here is a summary:
* `SamlObjectHandler` was renamed from `SamlRequestHandler` (It seems to be a good idea to me since its the base for both Request and Response handlers. But I can revert it if it does not sound right to you).
* `SamlResponseHandler` is extracted from `SamlAuthenticator` since many code can be reused by the new `SamlLogoutResponseHandler`

The class hierarchy is as follows:
```
SamlObjectHandler
├── SamlLogoutRequestHandler
└── SamlResponseHandler
    ├── SamlAuthenticator
    └── SamlLogoutResponseHandler
```

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlResponseHandlerTests.java:
```diff
@@ -0,0 +1,233 @@
+import static javax.xml.crypto.dsig.Transform.ENVELOPED;
+import static org.opensaml.saml.common.xml.SAMLConstants.SAML20_NS;
+
+public class SamlResponseHandlerTests extends SamlTestCase {
```
You are right. They are split with the IDE refactoring tool. So nothing should be really different.

ywangd committed:
Address feedback

ywangd committed:
Address feedback for removing auto-detect of redirect and post

ywangd commented:
Thanks a lot @jkakavas I have addressed your feedbacks. 

There seem to be some inconsistency in how we throw exceptions for saml handling. The method `SamlUtils#samlException(String msg, Object... args)` is used most often. I also used them in the new handler. This method implies `500` for HTTP response status, which does not feel right in many places. I am aware there is an overloaded version of this method where you can pass another cause exception for other status like `400`. I didn't use it mainly because it would be "inconsistent" with usages in many other places. I'd like to hear your opinions. Thanks

ywangd committed:
Fix test convention

ywangd committed:
checkstyle

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticatorTests.java:
```diff
@@ -108,97 +85,20 @@
 import static org.opensaml.saml.saml2.core.SubjectConfirmation.METHOD_BEARER;
 
-public class SamlAuthenticatorTests extends SamlTestCase {
+public class SamlAuthenticatorTests extends SamlResponseHandlerTests {
```
Apologies for the merge conflicts that the SamlAuthenticatorTests refactoring causes here

ywangd replied:
No worries. Wasn't too bad.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public void handle(boolean httpRedirect, String payload, Collection<String> allowedSamlRequestIds) {
+        final ParsedQueryString parsed = parseQueryStringAndValidateSignature(payload, "SAMLResponse");
```
Reading through the tests below, it looks like there is an implicit assumption that even when HTTP POST is used, we will receive `SAMLResponse=<Base64Endoded saml logout response here>` from Kibana and thus we could "parse" this as if it was a query string. I think we should be explicit and 

a) Declare that we expect kibana(or any caller) to pass the _value_ of the SAMLResponse POST parameter
b) Base64 decode this on the RestSamlLogoutAction
c) Operate on the raw bytes after that

similar to what we do for the SamlAuthenticateAction's parameter named content

ywangd replied:
The payload/data returned from a SAML idP with HTTP-Post binding uses `application/x-www-form-urlencoded` encoding, which takes the form of `SAMLResponse=<base64>&RelayState=xxxx`.  So it can indeed be processed as if it is a query string. 

But I think it is good point that we should keep this consistent with how SAMLReseponse is handled by SamlAuthenticationAction, for which I think Kibana performs the action of extracting the value of `SAMLResponse` parameter and passes it to ES so that we do not have to parse it (as query parameter) again. I will make the change and add a note in the code to clarify the expectation for Kibana.

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlCompleteLogoutRequest.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (queryString == null && content == null) {
```
use `Strings.hasText()` instead of null check ?

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandlerHttpRedirectTests.java:
```diff
@@ -0,0 +1,151 @@
+        assertThat(e.getMessage(), containsString("is not a 'success' response"));
+    }
+
+    public void testHandlerWillFailWhenUrlNotSigned() {
```
```suggestion
    public void testHandlerWillFailWhenLogoutResponseNotSigned() {
```

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandlerHttpRedirectTests.java:
```diff
@@ -0,0 +1,151 @@
+        assertThat(e.getMessage(), containsString("URL is not signed, but is required for HTTP-Redirect binding"));
+    }
+
+    public void testHandlerWillFailWhenUrlIsSignedButBindingIsHttpPost() throws URISyntaxException {
```
I don't think we need this

ywangd replied:
Removed

jkakavas reviewed (changes requested):
Added a few more comments

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java

Co-authored-by: Ioannis Kakavas <ikakavas@protonmail.com>

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,63 @@
+    public void handle(boolean httpRedirect, String payload, Collection<String> allowedSamlRequestIds) {
+        final ParsedQueryString parsed = parseQueryStringAndValidateSignature(payload, "SAMLResponse");
+        if (httpRedirect && parsed.hasSignature == false) {
+            throw samlException("URL is not signed, but is required for HTTP-Redirect binding");
```
You are right that it is not the entire URL that gets signed. But it is not just the encoded SAMLResponse either. It is rather a string taking the form of `SAMLResponse=xxx&RelayState=xxx&SigAlg=xx`. How about changing it to "Query string parameters are not signed, but is required for HTTP-Redirect binding"?

jkakavas replied:
The most important part here is for the user/administrator to realize what is wrong and why this failed. I think that "The SAML logout response is not signed" does that in a better way than "the URL is not signed" or "The query string is not signed" but I won't insist too much :)

ywangd committed:
Tidy up a bit due to latest change in master

ywangd committed:
Fix randomize of signer

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,63 @@
+
+        if (LOGOUT_RESPONSE_TAG_NAME.equals(root.getLocalName()) && SAML_NAMESPACE.equals(root.getNamespaceURI())) {
+            final LogoutResponse logoutResponse = buildXmlObject(root, LogoutResponse.class);
+            if (logoutResponse == null) {
```
Indeed `buildXmlObject` does not return `null`. I'll remove it. Just for discussion: Since Java does not have proper null safty, aggressively checking for `null` return value could be argued as defensive coding or invariant enforcing. I think this could be a reason that we have many `assert xxx` statement throughout the code base. The difference is that they do not run in production but helps to catch potential code change gaps in testing. What would be your take on this?

jkakavas replied:
We do have asserts throughout the code base and we do have places where we add extra null checks "just to be safe", but we have been inconsistent in doing so I think. I love defensive coding as the next person but I don't know where the line should be drawn so that we don't have a `if (x == null)` check every 5 lines. Thinking out loud, we should be able to trust the methods we write to not return null unless they explicitly document that and handle null checks in the method instead of everywhere we call that method

Tim had shared some thoughts on `assert`s here: https://github.com/elastic/elasticsearch/pull/52936#discussion_r386167255

ywangd replied:
Thanks. The difference is subtle but insightful.

ywangd committed:
Address feedback

ywangd commented:
Thanks @jkakavas, I have addressed your feedbacks. In particular, what data to take in for the HTTP-Post binding. I have follow the same pattern as in the Saml Authenticate case, i.e. the code expects a base64 encoded (but **not** URL encoded) SAMLResponse. The expectation is that Kibana (or other service) will prepare this string by extracting it from the HTTP-Post body. This change makes the handler code simpler (as you suggested).

Due to the refactor of `SamlAuthenticatorTests`, I add one more refactoring to work with the superclass (`SamlResponseHandlerTests`) extraction:
* Most content of `signResponseString` is extracted into a new superclass method called [`signSignableObject`](https://github.com/elastic/elasticsearch/pull/56316/files#diff-e411429aa33ed9aa06bcc1ffa59c5a92R213), which is reused to sign `Response`, `Assertion` and `LogoutResponse`. 
* Also fixed [this random signer setup](https://github.com/elastic/elasticsearch/pull/56316/files#diff-6cb21be4ddbab9f5569065cb108d6cbcR265) as I believe it was a typo to use `signResponse` twice.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-43264-saml-post-logout

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlCompleteLogoutRequest.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
```
could we also validate that the realm is not an empty string here so that we fail explicitly here instead of in the Transport action's `doExecute` ?

ywangd replied:
Added. Also removed the `@Nullable` annotation from the `realm` field.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlCompleteLogoutAction.java:
```diff
@@ -0,0 +1,62 @@
+    protected void doExecute(Task task, SamlCompleteLogoutRequest request, ActionListener<SamlCompleteLogoutResponse> listener) {
+        List<SamlRealm> realms = findSamlRealms(this.realms, request.getRealm(), null);
+        if (realms.isEmpty()) {
+            listener.onFailure(SamlUtils.samlException("Cannot find any matching realm for [{}]", request));
```
```suggestion
            listener.onFailure(SamlUtils.samlException("Cannot find any matching realm with name [{}]", request.getRealm()));
```

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlCompleteLogoutAction.java:
```diff
@@ -0,0 +1,62 @@
+        if (realms.isEmpty()) {
+            listener.onFailure(SamlUtils.samlException("Cannot find any matching realm for [{}]", request));
+        } else if (realms.size() > 1) {
+            listener.onFailure(SamlUtils.samlException("Found multiple matching realms [{}] for [{}]", realms, request));
```
Just a suggestion given that we search with a realm name only here. Traditionally we would search either with an ACS or a realm name that's why we had a more generic message here
```suggestion
            listener.onFailure(SamlUtils.samlException("Found multiple matching realms [{}] with name [{}]", realms, request.getRealm()));
```

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java:
```diff
@@ -0,0 +1,58 @@
+        if (LOGOUT_RESPONSE_TAG_NAME.equals(root.getLocalName()) && SAML_NAMESPACE.equals(root.getNamespaceURI())) {
+            final LogoutResponse logoutResponse = buildXmlObject(root, LogoutResponse.class);
+            // For HTTP-Redirect, we validate the signature while parsing the object from the query string
+            if (httpRedirect == false && logoutResponse.getSignature() == null) {
+                throw samlException("LogoutResponse is not signed, but a signature is required for HTTP-Post binding");
+            } else if (httpRedirect == false) {
+                validateSignature(logoutResponse.getSignature());
+            }
```
```suggestion
            if (httpRedirect == false) {
                if (logoutResponse.getSignature() == null) {
                    throw samlException("LogoutResponse is not signed, but a signature is required for HTTP-Post binding");
                }    
                validateSignature(logoutResponse.getSignature());
            }
```

jkakavas reviewed (approved):
LGTM Yang ,thanks for the iterations! I added some final comments and suggestions for your consideration but I think it's fine to merge this

Just a note, not to forget the API docs ( can do in a followup )

jkakavas commented:
> There seem to be some inconsistency in how we throw exceptions for saml handling. The method SamlUtils#samlException(String msg, Object... args) is used most often. I also used them in the new handler. This method implies 500 for HTTP response status, which does not feel right in many places. I am aware there is an overloaded version of this method where you can pass another cause exception for other status like 400. I didn't use it mainly because it would be "inconsistent" with usages in many other places. I'd like to hear your opinions. Thanks

The reason for `samlException()` is https://github.com/jkakavas/elasticsearch/blob/a7187ca208b4a7815d8be38c6214cecb610e3376/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java#L423 . If we're using it in other places and this affects the response and status code of our APIs, this is probably wrong and we should fix this. I'm not entirely certain what we should be returning in each case but we can start an issue and gather all places we throw such an Exception and see how we can "fix' them across the SAML APIs ( in a followup )

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlCompleteLogoutAction.java

Co-authored-by: Ioannis Kakavas <ikakavas@protonmail.com>

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java

Co-authored-by: Ioannis Kakavas <ikakavas@protonmail.com>

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlCompleteLogoutAction.java

Co-authored-by: Ioannis Kakavas <ikakavas@protonmail.com>

ywangd committed:
Address feedback

ywangd committed:
Add one more test for non-null realm field

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-43264-saml-post-logout

ywangd commented:
Thanks @jkakavas :)

> The reason for samlException() is https://github.com/jkakavas/elasticsearch/blob/a7187ca208b4a7815d8be38c6214cecb610e3376/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java#L423 . If we're using it in other places and this affects the response and status code of our APIs, this is probably wrong and we should fix this. I'm not entirely certain what we should be returning in each case but we can start an issue and gather all places we throw such an Exception and see how we can "fix' them across the SAML APIs ( in a followup )
>
The usage is indeed largely covered by the above exception check, which makes it sensible to use in `SamlAuthenticator` and `SamlRealm`. There are however many other places where it also gets used. I created #57331 to track potential improvement.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-43264-saml-post-logout

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlCompleteLogoutRequest.java:
```diff
@@ -0,0 +1,84 @@
+            validationException = addValidationError("queryString and content may not both be empty", validationException);
+        }
+        if (Strings.hasText(queryString) && Strings.hasText(content)) {
+            validationException = addValidationError("queryString and content may not both present", validationException);
```
What is our expectation for how a WebApp should behave if it gets a LogoutResponse over a HTTP-POST binding that also has URL parameters?

I think the rule for Kibana is:
- If the http method is `GET` send the queryString to Elasticsearch, and ignore (or reject) any body (which is very unlikely to exist)
- if the http method is `POST` ignore (or reject) the query parameters, and send the body to Elasticsearch.

Is that our intent?

ywangd replied:
A good question. To be honest, I haven't really considered this thoroughly. I think your suggestion makes sense because the code proceeds as HTTP-POST, e.g. base64 encode, no deflate and XML signature, when `content` is present. This basically says `content` implies `POST`, which in turn means `queryString` implies `GET`.

I would personally prefer to ignore the body for `GET` and query string for `POST` since the spec does not explicitly forbid them.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/saml/RestSamlCompleteLogoutAction.java:
```diff
@@ -0,0 +1,110 @@
+
+    private static final Logger logger = LogManager.getLogger(RestSamlCompleteLogoutAction.class);
+
+    static class Input {
```
Why do we have another class here, rather than just using `SamlCompleteLogoutRequest` ?

ywangd replied:
Because of copy/paste ... removed it. thanks

ywangd committed:
Address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-43264-saml-post-logout

ywangd commented:
@tvernum just ping you so this PR remains on your radar. thanks

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-43264-saml-post-logout
